linked_list/linked_list.py

```diff
@@ -9,71 +9,159 @@ def __init__(self, value, next_node = None):
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        if index < 0:
+            return None
+        
+        current_index = 0
+        current = self.head
+
+        if current is None:
+            return None
+
+        while current is not None and current_index < index:
+            current = current.next
+            current_index += 1
+        return current.value
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        if self.head is None:
+            return None
+
+        current = self.head
+
+        while current is not None:
+            if current.next:
+                current = current.next
+            else:
+                return current.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        if self.head is None:
+            new_node = Node(value)
+            self.head = new_node
+        else:
+            current = self.head
+
+            while current is not None:
+                if current.next:
+                    current = current.next
+                else:
+                    new_node = Node(value)
+                    current.next = new_node
+                    current = current.next
+                    break
+
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_max(self):
-        pass
+        if self.head is None:
+            return None
+        
+        current_max = self.head.value
+        current = self.head
+
+        while current is not None:
+            if current_max < current.value:
+                current_max = current.value
+            current = current.next
+        return current_max
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n); our worst case is that we need to traverse through the entire list to remove last element
+    # Space Complexity: O(1)
     def delete(self, value):
-        pass
+        if self.head is None:
+            return None
+
+        current = self.head
+        previous = self.head
+
+        while current is not None:
+            if current.value != value:
+                previous = current
+                current = current.next
+            else:
+                if self.head.value == value: # when we remove first element
+                    self.head = current.next
+                    current = current.next

```

Comment:
I would suggest moving this if outside the loop as you don't need to check to see if the head is what you're deleting with each iteration.